### PR TITLE
 feat(parachain/collator): handle AdvertiseCollation message received by a collator 

### LINKED ISSUE #3514: feat(parachain/collator): handle collation advertisements
## Issue summary

- This is part of collator protocol. 
- Handle both V1 and V2 messages
- Should be done after https://github.com/ChainSafe/gossamer/pull/3512 is merged.

- This message transfer is between two subsystems on a single node, so we have freedom on its structure.
- I have WIP code written for this, but not fully ready. Let me know if you want to take this up.

## Other information and links
Code reference in polkadot sdk
- v1 https://github.com/paritytech/polkadot-sdk/blob/f4827dcd16fc5585a8d15c6a32bc27ecde00e7f9/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L857

- v2 https://github.com/paritytech/polkadot-sdk/blob/f4827dcd16fc5585a8d15c6a32bc27ecde00e7f9/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L873

Make sure you follow the polkadot-sdk repo. The code has changed enough from the archived polkadot repo.

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues
Fixes #3514

## Primary Reviewer

@timwu20

### CODE REVIEW

File: dot/parachain/collator-protocol/backing_implicit_view.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package collatorprotocol
+
+import (
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// https://github.com/paritytech/polkadot-sdk/blob/d2fc1d7c91971e6e630a9db8cb627f8fdc91e8a4/polkadot/node/subsystem-util/src/backing_implicit_view.rs#L102
+type ImplicitView struct {
+}
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/backing_implicit_view.go:10]: add TODO and raise issue to implement it

File: dot/parachain/collator-protocol/backing_implicit_view.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// / for the leaf.
+// /
+// / `None` indicates that the block hash isn't part of the implicit view or that
+// / there are no known allowed relay parents.
+// /
+// / This always returns `Some` for active leaves or for blocks that previously
+// / were active leaves.
+// /
+// / This can return the empty slice, which indicates that no relay-parents are allowed
+// / for the para, e.g. if the para is not scheduled at the given block hash.
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/backing_implicit_view.go:28]: remove all extra `/`

File: dot/parachain/collator-protocol/message.go [hunk 3]
```diff
@@ -177,6 +178,225 @@ const (
...
+		CandidateHash:        candidateHash,
+		ParentHeadDataHash:   parentHeadDataHash,
+	}
+
+	responseChan := make(chan bool)
+
+	cpvs.SubSystemToOverseer <- struct {
+		responseChan     chan bool
+		canSecondRequest backing.CanSecond
+	}{
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message.go:210]: Could you create a struct type for this? We will need this structure in overseer.

File: dot/parachain/collator-protocol/message_test.go [hunk 4]
```diff
@@ -401,3 +401,406 @@ func TestHandleCollationMessageDeclare(t *testing.T) {
...
+
+	peerID := peer.ID("testPeerID")
+	testRelayParent := getDummyHash(5)
+	testParaID := parachaintypes.ParaID(5)
+
+	testCases := []struct {
+		description        string
+		advertiseCollation AdvertiseCollation
+		peerData           map[peer.ID]PeerData
+		// currentAssignments map[parachaintypes.ParaID]uint
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message_test.go:416]: Remove this comment, if you don't need it.

File: dot/parachain/collator-protocol/message_test.go [hunk 5]
```diff
@@ -401,3 +401,406 @@ func TestHandleCollationMessageDeclare(t *testing.T) {
...
+		perRelayParent map[common.Hash]PerRelayParent
+		net            Network
+		activeLeaves   map[common.Hash]ProspectiveParachainsMode
+		// success            bool
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message_test.go:420]: Remove this comment, if you don't need it.

File: dot/parachain/collator-protocol/message_test.go [hunk 6]
```diff
@@ -401,3 +401,406 @@ func TestHandleCollationMessageDeclare(t *testing.T) {
...
+		// 		ctrl := gomock.NewController(t)
+		// 		net := NewMockNetwork(ctrl)
+		// 		net.EXPECT().ReportPeer(peerset.ReputationChange{
+		// 			Value:  peerset.UnexpectedMessageValue,
+		// 			Reason: peerset.UnexpectedMessageReason,
+		// 		}, peerID)
+		// 		return net
+		// 	}(),
+		// 	activeLeaves: map[common.Hash]ProspectiveParachainsMode{},
+		// },
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message_test.go:557]: Remove this comment if you don't need it.

File: dot/parachain/collator-protocol/message_test.go [hunk 7]
```diff
@@ -401,3 +401,406 @@ func TestHandleCollationMessageDeclare(t *testing.T) {
...
+			activeLeaves: map[common.Hash]ProspectiveParachainsMode{
+				relayParent: {
+					isEnabled: false,
+				},
+			},
+			err: ErrDuplicateAdvertisement,
+		},
+		// {
+		// 	description: "fail with error peer limit reached",
+		// },
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message_test.go:729]: remove it or add todo

File: dot/parachain/collator-protocol/message.go [hunk 8]
```diff
@@ -177,6 +178,225 @@ const (
...
+	cpvs.fetchedCollations = append(cpvs.fetchedCollations, *collation)
+
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) handleAdvertisement(relayParent common.Hash, sender peer.ID,
+	prospectiveCandidate *ProspectiveCandidate) error {
+	// TODO:
+	// - tracks advertisements received and the source (peer id) of the advertisement
+
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message.go:299]: remove this Empty line
```suggestion
```

File: dot/parachain/collator-protocol/message.go [hunk 9]
```diff
@@ -291,7 +511,20 @@ func (cpvs CollatorProtocolValidatorSide) handleCollationMessage(
...
+			return propagate, errors.New("expected message to be advertise collation")
+		}
+
+		err := cpvs.handleAdvertisement(common.Hash(advertiseCollationMessage), sender, nil)
+		if err != nil {
+			return propagate, fmt.Errorf("handling v1 advertisement: %w", err)
+		}
+		// TODO:
+		// - tracks advertisements received and the source (peer id) of the advertisement
+
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message.go:525]: remove this Empty line
```suggestion
```

File: dot/parachain/collator-protocol/message.go (hunk 3 above)
Comment (kishansagathiya) [reply dot/parachain/collator-protocol/message.go:210]: This struct would be different for different messages. Type of request and response would be different most of the time.
I can create

File: dot/parachain/collator-protocol/message.go [hunk 10]
```diff
@@ -177,6 +178,224 @@ const (
...
+		logger.Debug("added collation to unfetched list")
+		collations.waitingQueue = append(collations.waitingQueue, UnfetchedCollation{
+			CollatorID:       collatorID,
+			PendingCollation: pendingCollation,
+		})
+	case Waiting:
+		// limit is not reached, it's allowed to second another collation
+		return cpvs.fetchCollation(pendingCollation)
+	case Seconded:
+		perRelayParent := cpvs.perRelayParent[relayParent]
```
[the remark this answers is not in the document]
Comment (kishansagathiya) [reply dot/parachain/collator-protocol/message.go:254]: That could be the case if we run multiple `handleAdvertisement` concurrently. We will surely do that. But we don't do that right now, so can't be a race condition.

File: dot/parachain/collator-protocol/message.go [hunk 11]
```diff
@@ -177,6 +178,227 @@ const (
...
+
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) handleAdvertisement(relayParent common.Hash, sender peer.ID,
+	prospectiveCandidate *ProspectiveCandidate) error {
+	// TODO:
+	// - tracks advertisements received and the source (peer id) of the advertisement
+	// - accept one advertisement per collator per source per relay-parent
+
```
Comment (edwardmack) [inline dot/parachain/collator-protocol/message.go:303]: Looking at the polkadot code I see that there is a signature verification that happens during this process: https://github.com/paritytech/polkadot-sdk/blob/f4827dcd16fc5585a8d15c6a32bc27ecde00e7f9/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L810C3-L810C91 

Do we need to do something like this as well?

**Review by edwardmack — APPROVED** [review]
Looks good to me, just a question regarding if/how we'll handle the signature validation.

File: dot/parachain/collator-protocol/message.go (hunk 11 above)
Comment (kishansagathiya) [reply dot/parachain/collator-protocol/message.go:303]: The link you have shared belongs to logic of processing `Declare` message, not `AdvertiseCollation` message.

In processing `Declare` message we have verified the signature at this permalink
https://github.com/ChainSafe/gossamer/blob/496560dd314d6095501935f9e632d70b1004af5f/dot/parachain/collator-protocol/message.go#L483

File: dot/parachain/collator-protocol/message.go [hunk 12]
```diff
@@ -177,6 +178,227 @@ const (
...
+	// TODO: return errors
+	pendingCollation := PendingCollation{
+		RelayParent:          relayParent,
+		ParaID:               paraID,
+		PeerID:               peerID,
+		ProspectiveCandidate: prospectiveCandidate,
+	}
+
+	switch collations.status {
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message.go:237]: add default case

File: dot/parachain/collator-protocol/validator_side.go [hunk 13]
```diff
@@ -228,6 +312,71 @@ type CollatorProtocolValidatorSide struct {
 	// Parachains we're currently assigned to. With async backing enabled
 	// this includes assignments from the implicit view.
 	currentAssignments map[parachaintypes.ParaID]uint
+
+	// state tracked per relay parent
+	perRelayParent map[common.Hash]PerRelayParent // map[replay parent]PerRelayParent
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/validator_side.go:317]: remove this extra comment
```suggestion
	perRelayParent map[common.Hash]PerRelayParent
```

**Review by axaysagathiya — APPROVED** [review]
LGTM

**Pull request merged**